### PR TITLE
Bilateral Filter

### LINKED ISSUE #702: [Feat] Bilateral filter
## 🚀 Feature

Would be great to have a differentiable version of Bilater Filtering `kornia.filter.bilateral_blur`

We can follow opencv signature:
https://docs.opencv.org/3.4/d4/d86/group__imgproc__filter.html#ga9d7064d478c95d60003cf839430737ed

Resource: http://jamesgregson.ca/bilateral-filtering-in-python.html

### PR DESCRIPTION
**Pull request by pikachu28**
### Description
I have added the algorithm for Bilateral filter using numpy as I am yet not comfortable with pytorch, closes [issue 702](https://github.com/kornia/kornia/issues/702#issue-718143682). Therefore, I ask reviewers to check the PR (@edgarriba ). I can do the same with pytorch but it will take me sometime to become comfortable with it. That is why I have put the PR on hold.

### Status
**Hold**

### Types of changes

- [x ] Breaking change (fix or new feature that would cause existing functionality to change)

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ x] Assign correct label
  - [x ] Assign PR to a reviewer
  - [ closed #702 x] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

File: kornia/filters/bilateral.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import math
+
+import numpy as np
+
+
+# gaussian filter
+def gfunc(x,y,sigma):
```
Comment (ducha-aiki) [inline kornia/filters/bilateral.py:7]: No need in implementing this, we have gaussian2d kernel https://github.com/kornia/kornia/blob/master/kornia/filters/kernels.py#L424

File: kornia/filters/bilateral.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    return (math.exp(-(x**2 + y**2)/(2*(sigma**2))))/(2*3.14*(sigma**2))
+
+def gaussFilter(size, sigma):
```
Comment (ducha-aiki) [inline kornia/filters/bilateral.py:10]: See above

**Comment by edgarriba** [comment]
@pikachu28 thanks for opening this PR (and nice username :) - Could you coordinate with @AnimeshMaheshwari22 , since he was also going for an implementation. Just one comment, interface ok , but the merge must be using pytorch since we do not accept other dependencies. Feel free to join our slack to discuss /cc @lferraz

File: kornia/filters/bilateral.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    out = np.zeros(size)
+    for i in range(size[0]):
+        for j in range(size[1]):
+            out[i,j] = gfunc(i-size[0]//2,j-size[1]//2, sigma )
+    return out/np.sum(out)
+
+def bfunc(i,j,fw,fh,image,sigma1, bilateralWFilter):
```
Comment (ducha-aiki) [inline kornia/filters/bilateral.py:17]: I believe, it would be better to implement the kernel (or couple of kernels) to be used with kornia filter2d
https://github.com/kornia/kornia/blob/master/kornia/filters/filter.py#L33

File: kornia/filters/bilateral.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            + '(kernel_size='
+            + str(self.kernel_size)
+            + ', '
+            + 'sigma='
+            + str(self.sigma)
+             + ')'
+        )
+    def forward(self, image):
+        "input image must be in numpy array"
+        return bilateral_blur(image, self.kernel_size, self.sigma)
```
Comment (ducha-aiki) [inline kornia/filters/bilateral.py:63]: It also would be goog to start with test cases (e.g. generate manually the input and desired output), or use OpenCV or skimage for the reference results

**Review by ducha-aiki — COMMENTED** [review]
Thank you for the nice PR! 
I put my comments in the lines.

File: kornia/filters/bilateral.py (hunk 2 above)
Comment (pikachu28) [reply kornia/filters/bilateral.py:10]: Okay, I will do as you suggested.

**Comment by pikachu28** [comment]
> @pikachu28 thanks for opening this PR (and nice username :) - Could you coordinate with @AnimeshMaheshwari22 , since he was also going for an implementation. Just one comment, interface ok , but the merge must be using pytorch since we do not accept other dependencies. Feel free to join our slack to discuss /cc @lferraz

Okay I will try to do the same with pytorch, I have joined the slack channel as well.

File: kornia/filters/bilateral.py (hunk 4 above)
Comment (AnimeshMaheshwari22) [reply kornia/filters/bilateral.py:63]: I have an implementation dependent only on PyTorch. The only doubt I have is wether the distance factor is to be used or not. I have already leveraged the pytorch parallel processing feature for faster implementation. 
Still I will add the distance factor to it.

**Comment by AnimeshMaheshwari22** [comment]
> @pikachu28 thanks for opening this PR (and nice username :) - Could you coordinate with @AnimeshMaheshwari22 , since he was also going for an implementation. Just one comment, interface ok , but the merge must be using pytorch since we do not accept other dependencies. Feel free to join our slack to discuss /cc @lferraz

I have an implementation dependent only on PyTorch. The only doubt I have is wether the distance factor is to be used or not. I have already leveraged the pytorch parallel processing feature for faster implementation.
Still I will add the distance factor to it.

File: kornia/filters/bilateral.py (hunk 1 above)
Comment (AnimeshMaheshwari22) [reply kornia/filters/bilateral.py:7]: > No need in implementing this, we have gaussian2d kernel https://github.com/kornia/kornia/blob/master/kornia/filters/kernels.py#L424

I have a doubt. The current gaussian kernel available with Kornia, is it distance dependent? Because we need 2 gaussian functions, intensity difference dependent and distance difference dependent. So I dont think we can use the current gaussian2d kernel. Because it does not factor in the difference of values.

Comment (ducha-aiki) [reply kornia/filters/bilateral.py:7]: At least this function does not depend on distance :) Other - yes

Comment (AnimeshMaheshwari22) [reply kornia/filters/bilateral.py:7]: > At least this function does not depend on distance :) Other - yes

I mean we have to calculate the gaussian of the distance of intensity the current pixel from the target pixel. The current kernel directly squares the values, as far as I understood the implementation.

Comment (edgarriba) [reply kornia/filters/bilateral.py:7]: this could be a potential separated functionality

File: kornia/filters/bilateral.py (hunk 3 above)
Comment (edgarriba) [reply kornia/filters/bilateral.py:17]: agree to have separated function for this

File: kornia/filters/bilateral.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        size = image.shape
+        fw, fh = kernel_size[0], kernel_size[1]
+        sigma1, sigma2 = sigma[0], sigma[1]
+        bilateral1 = 2*3.14*sigma2*sigma2*gaussFilter((fw,fh), sigma2)
+        if len(image.shape) < 3  or image.shape[2] == 1:
+            bilateralWFilter = np.resize(bilateral1,(*bilateral1.shape,1))
+        else:
+            bilateralWFilter = np.stack([bilateral1, bilateral1, bilateral1], axis=2)
+        out = np.zeros((size[0]-2*fw +1,size[1]-2*fh +1,size[2]))
+        for i in range(size[0]-2*fh +1):
```
Comment (edgarriba) [inline kornia/filters/bilateral.py:36]: we need to find a proper way to parallelize this for loops /cc @AnimeshMaheshwari22 @lferraz @ducha-aiki

File: kornia/filters/bilateral.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            for j in range(size[1]-2*fw +1):
+                out[i,j,:] = bfunc(i+fw-1, j+fh-1, fw, fh, image, sigma1, sigma2, bilateralWFilter)
+
+        if id == 1:
+            return np.resize(out, (out.shape[0], out.shape[1])).astype(np.uint8)
+        else:
+            return out.astype(np.uint8)
+
+class BilateralFilter():
```
Comment (edgarriba) [inline kornia/filters/bilateral.py:45]:
```suggestion
class BilateralFilter(nn.Module):
```

File: kornia/filters/bilateral.py (hunk 4 above)
Comment (edgarriba) [reply kornia/filters/bilateral.py:63]: use this as a reference: https://github.com/kornia/kornia/blob/master/test/filters/test_gaussian.py#L43 + include a regression test verifying let say opencv @AnimeshMaheshwari22 maybe you can help with that part to @pikachu28
